---
title: "matchedRNAATACAnalysis"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
atacMatrix <- readRDS("data/matchedATACMatrix.rds")
rnaMatrix <- readRDS("data/matchedRNAMatrix.rds")
atacResult <- readRDS("data/matchedATACResult.rds")
rnaResult <- readRDS("data/matchedRNAResult.rds")

ATACCoGAPS::cgapsPlot(atacResult, as.factor(colnames(atacMatrix)))
ATACCoGAPS::cgapsPlot(rnaResult, as.factor(colnames(rnaMatrix)))

ATACCoGAPS::heatmapPatternMatrix(atacResult, as.factor(colnames(atacMatrix)), col = viridis::magma(n = 9))
ATACCoGAPS::heatmapPatternMatrix(rnaResult, as.factor(colnames(rnaMatrix)), col = viridis::magma(n = 9))

```

```{r}
mouse = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
rna_seq_genes = biomaRt::getBM(attributes=c("mgi_symbol", "ensembl_gene_id"), mart=mouse)
saveRDS(rna_seq_genes, "data/ensemblIDToSymbol.rds")
# version used for paper saved in the data folder and can be loaded here to ensure reproducibility
# rna_seq_genes = readRDS("data/ensemblIDToSymbol.rds")

nosymbol = setdiff(rownames(rnaMatrix), rna_seq_genes[,2])
rminds = c()
for(i in 1:length(nosymbol)){
  ind = which(rownames(rnaMatrix) == nosymbol[i])
  rminds = c(rminds, ind)
}

rnaMatrix2 = rnaMatrix


for(i in 1:nrow(rnaMatrix2)) {
  ind = which(rna_seq_genes$ensembl_gene_id == rownames(rnaMatrix2)[i])
  if(length(ind) == 1){
  rownames(rnaMatrix2)[i] = rna_seq_genes$mgi_symbol[ind]
  }
}

motiflist = readRDS("data/ms_motiflist.rds")
tfInfo = readRDS("data/mouse_TF_info.rds")

atacGranges = ATACCoGAPS::peaksToGRanges(rownames(atacMatrix), "-")

#thresholding to save computational time, we have shown TFs found are robust to this thresholding (see Schep data analysis)
patternMotifs = ATACCoGAPS::motifPatternMatch(atacResult, atacGranges, motiflist, "mm10", 0.1)
TFs = ATACCoGAPS::getTFs(patternMotifs, tfInfo)

#load genes known to be regulated by TFs; downloaded from the TRRUST database
load("data/regInfo.rda")

TFGenes = ATACCoGAPS::findRegulatoryNetworks(TFs, mouseRegNets)

patMarkers = CoGAPS::patternMarkers(rnaResult)
genesRanks =patMarkers$PatternMarkerRanks
for(i in seq(nrow(genesRanks))) {
  ind = which(rna_seq_genes$ensembl_gene_id == rownames(genesRanks)[i])
  if(length(ind) == 1){
  rownames(genesRanks)[i] = rna_seq_genes$mgi_symbol[ind]
  }
}  

RNAvalidate <- ATACCoGAPS::RNAseqTFValidation(TFGenes = TFGenes, RNACoGAPSResult = genesRanks, RNAPatternSet = c(3,6), ATACPatternSet = c(1,4,6), matrix = TRUE)

#plot qqplot of p-values
pv.qqplot = function(pvector, main=NULL, ...) {
  o = -log10(sort(pvector,decreasing=F))
  e = -log10( 1:length(o)/length(o) )
  plot(e,o,pch=19,cex=1, main=main, ...,
       xlab=expression(Expected~~-log[10](italic(p))),
       ylab=expression(Observed~~-log[10](italic(p))),
       xlim=c(0,max(e)), ylim=c(0,max(o)))
  #45 degree line
  abline(0,1, col='red')
  #bonf. cutoff line
  abline(-log10(0.05/length(pvector)), 0)
}

pv.qqplot(RNAvalidate[[4]]$pval)

#check multiple testing significance
p.adjust(RNAvalidate[[4]]$pval, method = "BH") #only Tbx20 is MT-sig

#genes tbx20 is known to regulate
tbx20genes <- c("Nppa", "Nkx2-5", "Tbx2" ,"Mef2c")

#find ranks of tbx20 regulated genes as determined by PatternMarker statistic applied to RNA CoGAPS result
tbx20GeneRanks <- genesRanks[which(rownames(genesRanks) %in% tbx20genes), 6]


library(Mus.musculus)
tbx20GA <- ATACCoGAPS::geneAccessibility(tbx20genes, atacGranges, atacMatrix, Mus.musculus)

lapply(tbx20GA, rownames)

ATACCoGAPS::heatmapGeneAccessibility(tbx20GA$Mef2c, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(tbx20GA$`Nkx2-5`, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(tbx20GA$Tbx2, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))
heatmapGeneAccessibility(tbx20GA$Nppa, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))

lapply(tbx20GA, rownames)
```


```{r}
pat6GeneRanks <- genesRanks[,6]
topPat6Genes <- names(pat6GeneRanks[which(pat6GeneRanks %in% c(1:5))])

pat6GA <- ATACCoGAPS::geneAccessibility(topPat6Genes, atacGranges, atacMatrix, Mus.musculus)

ATACCoGAPS::heatmapGeneAccessibility(pat6GA$Klhl6, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(pat6GA$Kcng2, as.factor(colnames(atacMatrix)), colColors = c("darkorchid1", "blue"), col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(pat6GA$Obscn, as.factor(colnames(atacMatrix)))
ATACCoGAPS::heatmapGeneAccessibility(pat6GA$C330021F23Rik, as.factor(colnames(atacMatrix)))
ATACCoGAPS::heatmapGeneAccessibility(pat6GA$Cavin3, as.factor(colnames(atacMatrix)))
```

transfer learning with projectR
```{r}

geneRanges = function(txdb) {
  #get granges with gene id metadata
  g = GenomicFeatures::genes(txdb, columns="ENSEMBL")
  col = mcols(g)[["ENSEMBL"]]
  #create GRanges without metadata to make gene ids character vector
  genes = GenomicRanges::granges(g)[rep(seq_along(g), elementNROWS(col))]
  mcols(genes)[["ENSEMBL"]] = as.character(unlist(col))
  genes
}

library(Mus.musculus)

genes <- geneRanges(Mus.musculus)
ids <- mcols(genes)[["ENSEMBL"]]

#remove duplicates where the same ENSEMBL id is used for different genes
ids <- ids[-which(ids %in% ids[which(duplicated(ids))])]

overindsID <- which(ids %in% rownames(rnaMatrix))
overindsRNA <- which(rownames(rnaMatrix) %in% ids)
rnaMatrix3 <- rnaMatrix[overindsRNA,]
ids2 <- ids[overindsID]


orderInds = vector("numeric", length(ids2))
for(i in seq_along(rownames(rnaMatrix3))) {
  ind = which(ids2 == rownames(rnaMatrix3)[i])
  orderInds[i] = ind
  if(length(ind)>1){print(i)}
}

genes2 = genes[overinds]
genes2 = genes2[orderInds]



  olaps <- vector("list", length(genes2))
  toRemove <- NULL
  for(i in seq_along(genes2)) {
    olap <- GenomicRanges::findOverlaps(genes2[i], atacGranges,
                                        ignore.strand = TRUE)
    olapInds <- subjectHits(olap)
    if(length(olapInds) == 0) {
      print(paste("No overlapping peaks with", mcols(genes2)[["ENSEMBL"]][i]))
      toRemove <- c(toRemove, i)
    }
    else{
      olaps[[i]] <- olapInds
    }
  }

    
olaps[[1]]

#RNA to ATAC
olappingPeaks = unlist(olaps)
dupOP = olappingPeaks[which(duplicated(olappingPeaks))]
for(i in seq_along(olaps)){
  ol = olaps[[i]]
  if(length(which(ol %in% dupOP)) != 0) {
    if(length(ol[-which(ol %in% dupOP)]) == 0){
      olaps[[i]] = NA
    }
    else{
    olaps[[i]] = ol[-which(ol %in% dupOP)]
    }
  }
}

sum(duplicated(unlist(olaps))) #all are NA values for genes with no overlapping peaks

geneAggregateAtacMatrix = matrix(nrow = length(olaps), ncol = ncol(atacMatrix))
rownames(geneAggregateAtacMatrix) = mcols(genes2)[["ENSEMBL"]]

rowsToRemove = c()
for(i in seq(nrow(geneAggregateAtacMatrix))) {
  peaks = olaps[[i]]
  if(is.null(peaks)) {
    rowsToRemove = c(rowsToRemove, i)
  }
  else if(is.na(peaks)){
    rowsToRemove = c(rowsToRemove, i)
  }
  else if(length(peaks) == 1) {
    geneAggregateAtacMatrix[i,] = atacMatrix[peaks,]
  }
  else {
    agg = c()
    for(j in 2:length(peaks)) {
      if(j == 2){
        agg = unlist(atacMatrix[peaks[j-1],], use.names = F) + unlist(atacMatrix[peaks[j],], use.names = F)
      }
      else{
        agg = agg + atacMatrix[peaks[j],]
      }
    }
    geneAggregateAtacMatrix[i,] = agg
  }
}

geneAggregateAtacMatrix[1:5, 1:5]


geneAggATACMatrix2 = geneAggregateAtacMatrix[-rowsToRemove,]


#projectR
pR = projectR::projectR(geneAggATACMatrix2, rnaResult, rownames(geneAggATACMatrix2), rownames(rnaMatrix), full = T)

ATACCoGAPS::cgapsPlot(t(pR$projection), as.factor(colnames(atacMatrix)), matrix = T)
```



