---
title: "schep_main_pipeline"
output: html_document
editor_options: 
  chunk_output_type: console
---

Read in scATAC-seq data from GSE99172, Schep et al, 2017
```{r}
schepATAC <- read.csv(gzfile("data/GSE99172_scatac_counts.csv.gz"))

schepATAC[1:5, 1:5] #the first three columns contain peaks, the colnames are GSM numbers, and row one has cell names

#converting to numeric count matrix
countMatrix <- apply(as.matrix(schepATAC[-1, -c(1:3)]), 2, as.numeric)

#getting cell names
cellNames <- schepATAC[1, -c(1:3)]

#create vector of cell types
cellTypes <- unlist(lapply(cellNames, function(x){
  y <- stringr::str_split(x, "-")
  celltype = y[[1]][2]
  return(celltype)
}))

#getting peaks
peaks <- schepATAC[-1, c(1:3)]

#create GRanges from peaks
peakGranges <- GenomicRanges::makeGRangesFromDataFrame(peaks)

#create vector of peaks to use as labels
peakvec <- unlist(apply(peaks, 1, function(x){
  stringr::str_c(x[1], x[2], x[3], sep = "-")
}))

#naming cells and peaks
colnames(countMatrix) = cellTypes
rownames(countMatrix) = peakvec

countMatrix[1:5, 1:5]

#saving data
saveRDS(countMatrix, "data/SchepCountMatrix.rds")
saveRDS(peakGranges, "data/SchepPeakGranges.rds")
saveRDS(cellTypes, "data/SchepCellTypes.rds")
saveRDS(peakvec, "data/SchepPeaks.rds")
```


We now filter by sparsity to examine the results from different amounts of filtering.
```{r}
#perform filtering
filtered99 <- ATACCoGAPS::dataSubsetBySparsity(countMatrix, cellTypes, peakvec, cell_cut = 0.99, peak_cut = 0.99)

countMatrix99 <- filtered99$subset_data


filtered98 <- ATACCoGAPS::dataSubsetBySparsity(countMatrix, cellTypes, peakvec, cell_cut = 0.98, peak_cut = 0.98)

countMatrix98 <- filtered98$subset_data

#save data
saveRDS(countMatrix99, "data/SchepCountMatrix99Filtered.rds")
saveRDS(countMatrix98, "data/SchepCountMatrix98Filtered.rds")
```


Before running CoGAPS we create a parameters object. The number of patterns is set to 7 for all sets, running for 10000 iterations. The random seed ensures reproducibility of the factorization. We use singleCell and sparseoptimiztion mathods because we are using sparse single-cell data. We set the distributed parameter to genome-wide because their are more genomic features than cells. We run across one core per 10k peaks. 
```{r}
#parameters for full data set
params <- CogapsParams(nPatterns=7, nIterations=10000, seed=42, singleCell=TRUE, sparseOptimization=TRUE, distributed="genome-wide", geneNames=peakvec, sampleNames=cellTypes)
params <- setDistributedParams(params, nSets=12)
saveRDS(params, "data/7pat_10kIt_12sets_gw_params.rds")

#parameters for 99% sparsity filtered
params99 <- CogapsParams(nPatterns=7, nIterations=10000, seed=42, singleCell=TRUE, sparseOptimization=TRUE, distributed="genome-wide", geneNames=rownames(countMatrix99), sampleNames=colnames(countMatrix99))
params99 <- setDistributedParams(params99, nSets=9)
saveRDS(params99, "data/7pat_10kIt_9sets_gw_99sub_params.rds")

#parameters for 98% sparsity filtered
params98 <- CogapsParams(nPatterns=7, nIterations=10000, seed=42, singleCell=TRUE, sparseOptimization=TRUE, distributed="genome-wide", geneNames=rownames(countMatrix98), sampleNames=colnames(countMatrix98))
params98 <- setDistributedParams(params98, nSets=6)
saveRDS(params98, "data/7pat_10kIt_6sets_gw_98sub_params.rds")
```

We do not call CoGAPS locally and run the algorithm on AWS Batch servers, but to demonstrate:
```{r, eval = FALSE}
cogapsResult <- GWCoGAPS(data = countMatrix, params = params, nThreads = 12)

cogapsResult99 <- GWCoGAPS(data = countMatrix99, params = params99, nThreads = 9)

cogapsResult98 <- GWCoGAPS(data = countMatrix98, params = params98, nThreads = 6)
```


Read in results downloaded from AWS server run
```{r}
cgFullMatrix = readRDS("data/countMatrix_result.rds")
cg99Matrix = readRDS("data/countMatrix99_result.rds")
cg98Matrix = readRDS("data/countMatrix98_result.rds")
```

Plot the pattern matrix 
```{r}
#vector of colors to plot cell types by
col <- c("salmon", 'aquamarine', 'aquamarine3', "aquamarine4", "darkorchid1", 'limegreen', "magenta", "orangered", "darkgreen", "darkred", "red3", "darkorange")

#colors to plot patterns by
rowColors <- c("magenta", "darkorange", "aquamarine3", "salmon", "darkorchid1", "grey", "darkgreen")

#map factor of celltypes to more descriptive names
cellTypes <- plyr::mapvalues(as.factor(cellTypes), from = c("BJ","GM","GM12878","GM12878rep2","H1ESC","HL60","K562","LMPP","PB1022"   ,"SU070","SU353","TF1"), to = c("Fibroblasts", "GMLCL 1","GMLCL 2 Rep1","GMLCL 2 Rep2","H1ESCs","HL60 Leukemia","K562 Erythroleukemia","LMPP","Monocyte","AML Patient 070","AML Patient 353","TF1 Erythroblast"))

#get cell types for filtered data
cellTypes99 <- colnames(countMatrix99)
cellTypes99 <- plyr::mapvalues(as.factor(cellTypes99), from = c("BJ","GM","GM12878","GM12878rep2","H1ESC","HL60","K562","LMPP","PB1022"   ,"SU070","SU353","TF1"), to = c("Fibroblasts", "GMLCL 1","GMLCL 2 Rep1","GMLCL 2 Rep2","H1ESCs","HL60 Leukemia","K562 Erythroleukemia","LMPP","Monocyte","AML Patient 070","AML Patient 353","TF1 Erythroblast"))

#plot each pattern individually
ATACCoGAPS::cgapsPlot(cgFullMatrix, as.factor(cellTypes), cols= col, ylab = "Pattern Weight")

#plot all patterns together in a heatmap
ATACCoGAPS::heatmapPatternMatrix(cgFullMatrix, as.factor(cellTypes), cellCols = col, col = viridis::magma(n = 9))

#repeat for filtered data
ATACCoGAPS::cgapsPlot(cg99Matrix, cellTypes99, cols= col, ylab = "Pattern Weight")

ATACCoGAPS::heatmapPatternMatrix(cg99Matrix, cellTypes99, cellCols = col, rowColors = rowColors, col = viridis::magma(n=9))

ATACCoGAPS::cgapsPlot(cg98Matrix, as.factor(colnames(countMatrix98)), cols= col, ylab = "Pattern Weight")

ATACCoGAPS::heatmapPatternMatrix(cg98Matrix, as.factor(colnames(countMatrix98)), cellCols = col, col = viridis::magma(n = 9))
```
Filtering more than 99% sparse signal seems to remove cells that don't differentiate simply due to low read number. Filtering higher than that seems likely to just be losing useful signal.


We will continue with the 99% filtered data. 

To investigate differential accessibility we find the patternMarker peaks which differentiate each pattern. We can visualize these peaks in the original data.
```{r}
ATACCoGAPS::heatmapPatternMarkers(cgaps_result = cg99Matrix, atac_data = countMatrix99, celltypes = as.factor(colnames(countMatrix99)), numregions = 50, colColors = col, rowColors = c("magenta", "darkorange", "aquamarine3", "salmon", "darkorchid1", "limegreen", "darkgreen"), col = viridis::plasma(n = 2))


#make GRanges for subset peaks
subsetGranges <- ATACCoGAPS::peaksToGRanges(rownames(countMatrix99), sep= "-")

#get genes matched to patternMarker peaks
library(Homo.sapiens)
matchedGenes <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens)

#download pathways from msigDB
library(dplyr)
dPathways <- msigdbr::msigdbr(species = "Homo sapiens", category ="H") %>% dplyr::select(gs_name, gene_symbol) %>% as.data.frame()
#save current pathway set for reproducibility
saveRDS(dPathways, "data/msigdbrHallmarkPathways10_2019.rds")

#determine gene overlap between pathways and mathced genes
matchedPathways = ATACCoGAPS::pathwayMatch(matchedGenes, dPathways, p_threshold = 0.001)

#get summary table of pathways
pathwayTables = lapply(matchedPathways, function(x){x[4]})
names(pathwayTables) = unlist(lapply(c(1:7), function(x) {paste("Pattern", x)}))
pathwayTables

#saving 
saveRDS(pathwayTables, "data/PT.rds")

#comparing pathways found at different patternMarker score thresholds
matchedGenes0.03 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 0.03)
matchedPathways0.03 = ATACCoGAPS::pathwayMatch(matchedGenes0.03, dPathways, p_threshold = 0.001)

matchedGenes0.1 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 0.1)
matchedPathways0.1 = ATACCoGAPS::pathwayMatch(matchedGenes0.1, dPathways, p_threshold = 0.001)

matchedGenes0.3 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 0.3)
matchedPathways0.3 = ATACCoGAPS::pathwayMatch(matchedGenes0.3, dPathways, p_threshold = 0.001)

matchedGenes0.5 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 0.5)
matchedPathways0.5 = ATACCoGAPS::pathwayMatch(matchedGenes0.5, dPathways, p_threshold = 0.001)

matchedGenes0.8 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 0.8)
matchedPathways0.8 = ATACCoGAPS::pathwayMatch(matchedGenes0.8, dPathways, p_threshold = 0.001)

matchedGenes1 <- ATACCoGAPS::genePatternMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, genome = Homo.sapiens, scoreThreshold = 1)
matchedPathways1 = ATACCoGAPS::pathwayMatch(matchedGenes1, dPathways, p_threshold = 0.001)

lapply(matchedPathways0.03, function(x){x[4]})
lapply(matchedPathways0.1, function(x){x[4]})
lapply(matchedPathways0.3, function(x){x[4]})
lapply(matchedPathways0.5, function(x){x[4]})
lapply(matchedPathways0.8, function(x){x[4]})
lapply(matchedPathways1, function(x){x[4]})

#the pathways identified at low thresholds are found as the most significant at higher thresholds; relatively robust to choice of threshold value, though more pathways appear when higher thresholds are used
```



Find potential TF bindings
```{r}
#match motifs to peaks and find TF binding for those peaks
motifTFResults <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19")

#get summary results
mtfResults = motifTFResults$tfMatchSummary
names(mtfResults) = unlist(lapply(c(1:7), function(x) {paste("Pattern", x)}))
mtfResults
#save motifTF enrichment
saveRDS(mtfResults, "data/motifTFResults.rds")

#other score thresholds
motifTFResults0.1 <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19", scoreThreshold = 0.1)
motifTFResults0.3 <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19", scoreThreshold = 0.3)
motifTFResults1 <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cg99Matrix, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19", scoreThreshold = 1)

motifTFResults0.1$tfMatchSummary
motifTFResults0.3$tfMatchSummary
motifTFResults1$tfMatchSummary
#motif results are quite robust across different thresholds

#summary for most enriched TF binding in pattern7 (monocyte)
motifTFResults$tfDescriptions[[7]][which(motifTFResults$tfDescriptions[[7]][,2]=="IRF1"), 1]


#check accessibility of predicted TFs for K562 cells
KTFAcc <- ATACCoGAPS::geneAccessibility(geneList = c("TAL1", "EGR1", "RREB1", "NFE2"), peakGranges = subsetGranges, atacData = countMatrix99, genome = Homo.sapiens)


mean(apply(countMatrix99, 2, sum)) #about 6600 reads/cell
binaryMatrix <- (countMatrix99 > 0) + 0
averageAcc <- mean(apply(binaryMatrix, 1, sum))/ncol(countMatrix99) #about 5-6% of cells are accessible for a given peak

#find the amount of times cells are accessible in peaks overlapping with genes of interest. To produce rough estimate of accessibility to compare to the peak average for that cell type

lapply(KTFAcc, ATACCoGAPS::foldAccessibility, cellTypes99, "K562 Erythroleukemia", binaryMatrix)

#plot accessibility
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = KTFAcc$EGR1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = KTFAcc$NFE2, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = KTFAcc$RREB1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = KTFAcc$TAL1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))

#check accessibility of predicted TFs for monocytes
mnTFAcc <- ATACCoGAPS::geneAccessibility(geneList = c("IRF1", "CEBPA", "SPI1", "STAT1"), peakGranges = subsetGranges, atacData = countMatrix99, genome = Homo.sapiens)
#plot accessibility
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = mnTFAcc$SPI1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = mnTFAcc$STAT1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = mnTFAcc$CEBPA, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))
ATACCoGAPS::heatmapGeneAccessibility(genePeaks = mnTFAcc$IRF1, celltypes = cellTypes99, colColors = col, col = c("white", "darkgreen"))

lapply(mnTFAcc, ATACCoGAPS::foldAccessibility, cellTypes99, "Monocyte", binaryMatrix)
```


Classifying cells using the pattern matrix and comparing to ground truth
```{r}
#number of each cell type
summary(as.factor(colnames(countMatrix99)))

#get patternMarker statistic for the Pattern Matrix
pm <- CoGAPS::patternMarkers(cg99Matrix, axis = 2)
patMarkers <- pm$PatternMarkers

lapply(patMarkers, length)


#apply pattern Marker cell classifications
pmcc <- ATACCoGAPS::patternMarkerCellClassifier(cg99Matrix)


pMatrix <- pmcc$predictionMatrix
cellClass <- pmcc$cellClassifier

cellTypes <- colnames(countMatrix99)

#determine the cell line most significantly associated with each pattern using a wilcoxon rank sum test
#get PatternMatrix
patternMatrix = getSampleFactors(cg99Matrix)
#paired wilcoxon rank sum test
apply(patternMatrix, 2, pairwise.wilcox.test, g = cellTypes)
#plot patterns again for refernece
ATACCoGAPS::cgapsPlot(cg99Matrix, cellTypes, col)

#test classifications of significant cell types to pattern by finding ROC AUC for each cell line
Kclass <- rep(0, length(cellTypes))
Kclass[which(cellTypes == "K562")] <- 1
Kroc <- pROC::roc(Kclass, pMatrix[,1])
pROC::auc(Kroc)

TFclass <- rep(0, length(cellTypes))
TFclass[which(cellTypes == "TF1")] <- 1
TFroc <- pROC::roc(TFclass, pMatrix[,2])
pROC::auc(TFroc)

GMclass <- rep(0, length(cellTypes))
GMclass[c(which(cellTypes=="GM"), which(cellTypes=="GM12878"),
                             which(cellTypes=="GM12878rep2"))] <- 1

GMroc <- pROC::roc(GMclass, pMatrix[,3])
pROC::auc(GMroc)


BJclass <- rep(0, length(cellTypes))
BJclass[which(cellTypes == "BJ")] <- 1
BJroc <- pROC::roc(BJclass, pMatrix[,4])
pROC::auc(BJroc)

H1class <- rep(0, length(cellTypes))
H1class[which(cellTypes == "H1ESC")] <- 1
H1roc <- pROC::roc(H1class, pMatrix[,5])
pROC::auc(H1roc)

HLclass <- rep(0, length(cellTypes))
HLclass[which(cellTypes == "HL60")] <- 1
HLroc <- pROC::roc(HLclass, pMatrix[,6])
pROC::auc(HLroc)

PBclass <- rep(0, length(cellTypes))
PBclass[which(cellTypes == "PB1022")] <- 1
PBroc <- pROC::roc(PBclass, pMatrix[,7])
pROC::auc(PBroc)

rocList <- list(Kroc, TFroc, GMroc, BJroc, H1roc, HLroc, PBroc)


aucVals <- unlist(lapply(rocList, pROC::auc))
CellFct <- as.factor(seq(length(aucVals)))
CellFct <- plyr::mapvalues(CellFct, from = as.character(CellFct), to = c("K562 Erythroleukemia", "TF1 Erythroblast", "GM LCLS", "Fibroblasts", "H1ESCs", "HL60 Leukemia", "Monocyte"))

#plot AUC vals
plotData <- data.frame(AUC = aucVals, Pattern = seq(length(aucVals)), AnnotatedCellType = CellFct)

library(ggplot2)
ggplot(plotData, aes(x = Pattern, y = AUC)) +
  geom_point(aes(col = AnnotatedCellType), size = 4) +
  scale_colour_manual(values = c("magenta", "darkorange", "aquamarine3", "salmon",      "darkorchid1", "grey", "darkgreen")) +
  ylim(c(0,1)) +
  scale_x_continuous(breaks = seq(1, 7, 1))

col2 <- c("black", RColorBrewer::brewer.pal(7, "Dark2"))

cellClass <- plyr::mapvalues(as.factor(cellClass), from = c("0", "1", "2", "3", "4", "5", "6", "7"), to = c("Unclassified", "Pattern1", "Pattern2", "Pattern3", "Pattern4", "Pattern5", "Pattern6", "Pattern7"))

ATACCoGAPS::cgapsPlot(cg99Matrix, cellClass, col2)

rowColors = col2[2:8]

#plot clustered by Pattern class
ATACCoGAPS::heatmapPatternMatrix(cg99Matrix, as.factor(cellClass), cellCols = col2, rowColors = rowColors, col = viridis::magma(n=9))

#plot PatternMarkers by pattern class
ATACCoGAPS::heatmapPatternMarkers(cgaps_result = cg99Matrix, atac_data = countMatrix99, celltypes = as.factor(cellClass), numregions = 50, colColors = col, rowColors = rowColors, col = viridis::plasma(n = 2))

#getting order of celltypes matched with order of pattern classifications for plotting
cfct <- factor(colnames(countMatrix99), levels = c("K562", "TF1", "GM", "GM12878", "GM12878rep2", "BJ", "H1ESC", "HL60", "LMPP", "PB1022", "SU070", "SU353"), ordered = TRUE)

cfct2 <- plyr::mapvalues(as.factor(cfct), from = c("K562", "TF1", "GM", "GM12878", "GM12878rep2", "BJ", "H1ESC", "HL60", "LMPP", "PB1022", "SU070", "SU353"), to = c("K562 Erythroleukemia", "TF1 Erythroblast", "GM LCL 1","GM LCL 2 Rep1","GM LCL 2 Rep2", "BJ Fibroblasts", "H1ESCs","HL60 Leukemia", "LMPP", "PB1022 Monocyte", "AML Patient 070", "AML Patient 353"))

combinedClasses <- data.frame(cellClasses = as.factor(cellClass), cellType = cfct2)
sortClasses <- combinedClasses[order(combinedClasses$cellClasses),]

col <- c("magenta", "darkorange", 'aquamarine', 'aquamarine3', "aquamarine4", "salmon", "darkorchid1", 'limegreen', "orangered", "darkgreen", "darkred", "red3")

#plot heatmap of the pattern matrix ordered by PatternMarker classfication
ATACCoGAPS::heatmapPatternMatrix(cg99Matrix, sortClasses$cellType, cellCols = col, col = viridis::magma(n=9), sort = FALSE)

#plot individual patterns
ATACCoGAPS::cgapsPlot(cg99Matrix, cfct2, col)

#pattern 7 with LMPP and AML included
pat7class <- rep(0, length(cellTypes))
pat7class[c(which(cellTypes == "PB1022"), which(cellTypes == "LMPP"), which(cellTypes == "SU070"), which(cellTypes == "SU353"))] <- 1
pat7roc <- pROC::roc(pat7class, pMatrix[,7])
pROC::auc(pat7roc)

#adjusted Rand Index for single evaluation of clustering accuracy
#convert cell types to numeric values based on pattern association
cellTypes[which(cellTypes == "K562")] <- 1
cellTypes[which(cellTypes == "TF1")] <- 2
cellTypes[c(which(cellTypes=="GM"), which(cellTypes=="GM12878"),
                             which(cellTypes=="GM12878rep2"))] <- 3
cellTypes[which(cellTypes == "BJ")] <- 4
cellTypes[which(cellTypes == "H1ESC")] <- 5
cellTypes[which(cellTypes == "HL60")] <- 6
cellTypes[which(cellTypes == "PB1022")] <- 7

#get ARI
mclust::adjustedRandIndex(cellClass, cellTypes)

#if all celltypes most strongly associated with pattern 7 are assigned to pattern 7 
cellTypes[c(which(cellTypes=="LMPP"), which(cellTypes=="SU070"),
                             which(cellTypes=="SU353"))] <- 7

mclust::adjustedRandIndex(cellClass, cellTypes)

```


projectR for projection of patterns of accessibility into other data sets (hematopoietic cells from Buenrostro et al, 2018, GSE96769)
```{r}
#getting count matrix - peaks x cells
repmis::source_data("https://github.com/FertigLab/ATACCoGAPS/blob/master/BuenrostroFinalSubsetData.Rdata?raw=true")
#getting GRanges for peaks
repmis::source_data("https://github.com/FertigLab/ATACCoGAPS/blob/master/BuenrostroGRanges.Rdata?raw=true")
#getting celltypes
repmis::source_data("https://github.com/FertigLab/ATACCoGAPS/blob/master/BuenrostroCellTypes.Rdata?raw=true")

#run projectR
projectRResults <- ATACCoGAPS::ATACTransferLearning(newData = BuenrostroFinalSubsetData, CoGAPSResult = cg99Matrix, originalPeaks = rownames(countMatrix99), originalGranges = subsetGranges,newGranges = BuenrostroGRanges)


#plot the output patterns
set.seed(22)
BuenColors <- gtools::permute(rainbow(10))
BuenColors[7] <- "darkgreen"

ATACCoGAPS::cgapsPlot(t(projectRResults$projection), as.factor(BuenrostroCellTypes), matrix = TRUE, col = BuenColors, ylab = "Pattern Weight")

#testing monocyte significance in Buenrostro data for monocyte pattern from Schep data
pairwise.wilcox.test(projectRResults$projection[7,], BuenrostroCellTypes, p.adjust.method = "BH")

pRPvals <- projectRResults$pval

#monocytes are very significant in projectR transfer, and there is generally increased significance for an immune cell type
mean(pRPvals[7,])
mean(pRPvals[7, which(BuenrostroCellTypes == "mono")])

#as compared to the fibroblast pattern
mean(pRPvals[4,])

#B cell derived LCL pattern 
pairwise.wilcox.test(projectRResults$projection[3,], BuenrostroCellTypes, p.adjust.method = "BH")
mean(pRPvals[3, which(BuenrostroCellTypes == "CLP")])
mean(pRPvals[3, which(BuenrostroCellTypes == "pDC")])


#Erythroleukia pattern
pairwise.wilcox.test(projectRResults$projection[1,], BuenrostroCellTypes, p.adjust.method = "BH")
mean(pRPvals[1, which(BuenrostroCellTypes == "MEP")])

#boxplot of patterns for pattern weight by cell type; helps visualize the difference in association to the pattern 

pr <- projectRResults$projection
p1 <- pr[1,]
boxplot(p1 ~ as.factor(BuenrostroCellTypes), xlab = "Cell Type", ylab = "Pattern Weight")

p7 <- pr[7,]
boxplot(p7 ~ as.factor(BuenrostroCellTypes), xlab = "Cell Type", ylab = "Pattern Weight")
```


Motif Summarization
```{r}
#motif list of PWM files downloaded from cisBP
motifList <- readRDS("data/motiflist")

#match to motifs
motifsSchep <- ATACCoGAPS::motifSummarization(motifList = motifList, scATACData = countMatrix99, granges = subsetGranges, genome = "hg19", cellNames = colnames(countMatrix99), pCutoff = 5e-09)

#convert to log space, as some motifs have high counts
logMotifs <- log1p(motifsSchep)

#CoGAPS parameters
params <- CogapsParams(nPatterns=9, nIterations=10000, seed=42, singleCell=TRUE, sparseOptimization=TRUE, distributed="genome-wide", geneNames=TFBSTools::ID(motifList), sampleNames=colnames(countMatrix99))
params <- setDistributedParams(params, nSets=2)

saveRDS(params, file="data/motifs_9pat_10kIt_4sets.rds")

#load CoGAPS result; algorithm was run on AWS cloud
motifResult = readRDS("data/SchepByMotifsResult.rds")

#plot pattern MAtrix
ATACCoGAPS::cgapsPlot(motifResult, as.factor(colnames(countMatrix99)), col, ylab = "Pattern Weight")
ATACCoGAPS::heatmapPatternMatrix(motifResult, as.factor(colnames(countMatrix99)), cellCols = col, col = viridis::magma(n=9))

#get patternMarker motifs
patternMarkerMotifs <- CoGAPS::patternMarkers(motifResult)
motifPatList <- patternMarkerMotifs$PatternMarkers

#find TFs corresponding to patternMarker motifs
TFinfo <- readRDS("data/TF_Info_all_motifs.rds") 
TFs <- ATACCoGAPS::getTFs(motifPatList, TFinfo)

#pattern 6 has strong signal for fibroblasts
pat6MarkerMotifs <- motifPatList[[6]]

#finding motifs in same motif list for cogaps ran by peaks
cMotifMatch <- ATACCoGAPS::motifPatternMatch(cg99Matrix, subsetGranges, motifList, "hg19", scoreThreshold = 0.05)

length(cMotifMatch[[4]])
length(pat6MarkerMotifs)
#compare the motifs found for the two fibroblast-associated patterns
intersect(pat6MarkerMotifs, cMotifMatch[[4]])

#pattern4 has strong signal for GM LCLs
pat4MarkerMotifs <- motifPatList[[4]]
length(pat4MarkerMotifs)
intersect(pat4MarkerMotifs, cMotifMatch[[3]])

pairwise.wilcox.test(getSampleFactors(motifResult)[,2], as.factor(colnames(countMatrix99)))#pattern2 is most strongly associated to H1ESCs
pat2MarkerMotifs <- motifPatList[[2]]
length(pat2MarkerMotifs)
intersect(pat2MarkerMotifs, cMotifMatch[[5]])

#we find different motifs with the different summarizations; indicates different regulatory patterns discovered
```

Higher number of patterns
```{r}
#cogaps run for 18 patterns
cgapsResult18Pat <- readRDS("data/SchepCoGAPSResult18Pat.rds")

#plot patternMatrix 
ATACCoGAPS::heatmapPatternMatrix(cgapsResult18Pat, as.factor(colnames(countMatrix99)), cellCols = col, col = viridis::magma(n=9))

#find associated pathways
matchedGenes <- ATACCoGAPS::genePatternMatch(cogapsResult = cgapsResult18Pat, generanges = subsetGranges, genome = Homo.sapiens)
matchedpaths <- ATACCoGAPS::pathwayMatch(matchedGenes, dPathways, p_threshold = 0.001)
lapply(matchedpaths, function(x){x[[4]]})

#find associated TFs
motifTFResults <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cgapsResult18Pat, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19")

motifTFResults$tfMatchSummary

#cogaps run for 13 patterns
cgapsResult13Pat <- readRDS("data/SchepCoGAPSResult13Pat.rds")
ATACCoGAPS::heatmapPatternMatrix(cgapsResult13Pat, as.factor(colnames(countMatrix99)), cellCols = col, col = viridis::magma(n=9))

matchedGenes <- ATACCoGAPS::genePatternMatch(cogapsResult = cgapsResult13Pat, generanges = subsetGranges, genome = Homo.sapiens)
matchedpaths <- ATACCoGAPS::pathwayMatch(matchedGenes, dPathways, p_threshold = 0.001)
lapply(matchedpaths, function(x){x[[4]]})

motifTFResults <- ATACCoGAPS::simpleMotifTFMatch(cogapsResult = cgapsResult13Pat, generanges = subsetGranges, organism = "Homo sapiens", genome = "hg19")

motifTFResults$tfMatchSummary

#ARI
pmcc <- ATACCoGAPS::patternMarkerCellClassifier(cgapsResult13Pat)

pMatrix <- pmcc$predictionMatrix
cellClass <- pmcc$cellClassifier

cellTypes <- colnames(countMatrix99)

cellTypes[which(cellTypes == "PB1022")] <- 1
cellTypes[which(cellTypes == "H1ESC")] <- 2
cellTypes[which(cellTypes == "K562")] <- 3
cellTypes[which(cellTypes == "TF1")] <- 5
cellTypes[which(cellTypes == "BJ")] <- 7
cellTypes[which(cellTypes == "HL60")] <- 9
cellTypes[which(cellTypes == "LMPP")] <- 10
cellTypes[c(which(cellTypes == "GM"), which(cellTypes == "GM12878"),
            which(cellTypes == "GM12878rep2"))] <- 12

mclust::adjustedRandIndex(cellClass, cellTypes)
```
Higher pattern numbers are able to distinguish LMPP cells and monocytes a bit more. Regulatory programs identified are generally robust to the change in pattern number; the patterns that seem to distinguish the same celltypes as lower pattern numbers learn the same pathways and TFs.
